File: swift/Sources/DSA/DataStructures/SinglyLinkedList.swift
/// Singly linked list with forward traversal.
/// Time: O(1) for prepend, O(n) for other operations
/// Space: O(n)
public struct SinglyLinkedList<T> {
    private class Node {
        var value: T
        var next: Node?

        init(_ value: T, next: Node? = nil) {
            self.value = value
            self.next = next
        }
    }

    private var head: Node?
    private var size = 0

    public init() {}

    public mutating func prepend(_ value: T) {
        head = Node(value, next: head)
        size += 1
    }

    public mutating func append(_ value: T) {
        let node = Node(value)
        if head == nil {
            head = node
        } else {
            var current = head
            while current?.next != nil {
                current = current?.next
            }
            current?.next = node
        }
        size += 1
    }

    public mutating func insert(_ index: Int, _ value: T) -> Bool {
        guard index >= 0 && index <= size else { return false }
        if index == 0 {
            prepend(value)
            return true
        }
        var current = head
        for _ in 0..<(index - 1) {
            current = current?.next
        }
        current?.next = Node(value, next: current?.next)
        size += 1
        return true
    }

    public mutating func removeAt(_ index: Int) -> T? {
        guard index >= 0 && index < size else { return nil }
        if index == 0 {
            let value = head?.value
            head = head?.next
            size -= 1
            return value
        }
        var current = head
        for _ in 0..<(index - 1) {
            current = current?.next
        }
        let value = current?.next?.value
        current?.next = current?.next?.next
        size -= 1
        return value
    }

    public func get(_ index: Int) -> T? {
        guard index >= 0 && index < size else { return nil }
        var current = head
        for _ in 0..<index {
            current = current?.next
        }
        return current?.value
    }

    public func indexOf(_ value: T) -> Int where T: Equatable {
        var current = head
        var index = 0
        while let node = current {
            if node.value == value { return index }
            current = node.next
            index += 1
        }
        return -1
    }

    public func contains(_ value: T) -> Bool where T: Equatable {
        indexOf(value) != -1
    }

    public var isEmpty: Bool {
        size == 0
    }

    public var count: Int {
        size
    }

    public mutating func clear() {
        head = nil
        size = 0
    }

    public func toArray() -> [T] {
        var result: [T] = []
        var current = head
        while let node = current {
            result.append(node.value)
            current = node.next
        }
        return result
    }
}

File: rust/src/data_structures/singly_linked_list.rs
/// A singly linked list implementation.
/// Time: O(1) for push_front, O(n) for push_back, search
/// Space: O(n)

type Link<T> = Option<Box<Node<T>>>;

#[derive(Debug)]
struct Node<T> {
    value: T,
    next: Link<T>,
}

#[derive(Debug)]
pub struct SinglyLinkedList<T> {
    head: Link<T>,
    len: usize,
}

impl<T> SinglyLinkedList<T> {
    pub fn new() -> Self {
        SinglyLinkedList { head: None, len: 0 }
    }

    pub fn push_front(&mut self, value: T) {
        let new_node = Box::new(Node {
            value,
            next: self.head.take(),
        });
        self.head = Some(new_node);
        self.len += 1;
    }

    pub fn pop_front(&mut self) -> Option<T> {
        self.head.take().map(|node| {
            self.head = node.next;
            self.len -= 1;
            node.value
        })
    }

    pub fn push_back(&mut self, value: T) {
        let new_node = Box::new(Node { value, next: None });

        if self.head.is_none() {
            self.head = Some(new_node);
        } else {
            let mut current = self.head.as_mut();
            while let Some(node) = current {
                if node.next.is_none() {
                    node.next = Some(new_node);
                    break;
                }
                current = node.next.as_mut();
            }
        }
        self.len += 1;
    }

    pub fn peek_front(&self) -> Option<&T> {
        self.head.as_ref().map(|node| &node.value)
    }

    pub fn peek_front_mut(&mut self) -> Option<&mut T> {
        self.head.as_mut().map(|node| &mut node.value)
    }

    pub fn len(&self) -> usize {
        self.len
    }

    pub fn is_empty(&self) -> bool {
        self.head.is_none()
    }

    pub fn clear(&mut self) {
        self.head = None;
        self.len = 0;
    }

    pub fn contains(&self, value: &T) -> bool
    where
        T: PartialEq,
    {
        let mut current = self.head.as_ref();
        while let Some(node) = current {
            if &node.value == value {
                return true;
            }
            current = node.next.as_ref();
        }
        false
    }
}

impl<T> Default for SinglyLinkedList<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T> Drop for SinglyLinkedList<T> {
    fn drop(&mut self) {
        let mut current = self.head.take();
        while let Some(mut node) = current {
            current = node.next.take();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_push_pop_front() {
        let mut list = SinglyLinkedList::new();
        list.push_front(1);
        list.push_front(2);
        list.push_front(3);

        assert_eq!(list.pop_front(), Some(3));
        assert_eq!(list.pop_front(), Some(2));
        assert_eq!(list.pop_front(), Some(1));
        assert_eq!(list.pop_front(), None);
    }

    #[test]
    fn test_push_back() {
        let mut list = SinglyLinkedList::new();
        list.push_back(1);
        list.push_back(2);
        list.push_back(3);

        assert_eq!(list.pop_front(), Some(1));
        assert_eq!(list.pop_front(), Some(2));
    }

    #[test]
    fn test_peek() {
        let mut list = SinglyLinkedList::new();
        assert_eq!(list.peek_front(), None);

        list.push_front(42);
        assert_eq!(list.peek_front(), Some(&42));
    }

    #[test]
    fn test_contains() {
        let mut list = SinglyLinkedList::new();
        list.push_back(1);
        list.push_back(2);
        list.push_back(3);

        assert!(list.contains(&2));
        assert!(!list.contains(&5));
    }

    #[test]
    fn test_len() {
        let mut list = SinglyLinkedList::new();
        assert_eq!(list.len(), 0);

        list.push_front(1);
        list.push_front(2);
        assert_eq!(list.len(), 2);

        list.pop_front();
        assert_eq!(list.len(), 1);
    }
}
